### PR TITLE
new: usr: #19: Support storing catalog in a DB

### PR DESCRIPTION
**Pull request by adeshr**

### CODE REVIEW

File: jdbc/pom.xml [hunk 1]
```diff
@@ -105,6 +105,16 @@
             <artifactId>jcl-over-slf4j</artifactId>
             <version>1.7.13</version>
         </dependency>
+        <dependency>
+            <groupId>org.jdbi</groupId>
```
Comment (vrajat) [inline jdbc/pom.xml]: Can you investigate if encryption is supported ? Similar to AES encryption of fields through Hibernate

**Comment by vrajat** [comment]
Have you forgotten to add the flyway migration files ?

File: jdbc/src/main/java/com/qubole/quark/jdbc/QuarkDriver.java [hunk 2]
```diff
@@ -101,7 +101,10 @@ public Meta createMeta(AvaticaConnection connection) {
         ((QuarkConnectionImpl) connection).getProperties());
   }
 
-  public String getSchemaFactoryPath() {
+  public String getSchemaFactoryPath(boolean isDbSchema) {
```
Comment (vrajat) [inline jdbc/src/main/java/com/qubole/quark/jdbc/QuarkDriver.java]: This needs to be cleaned up. I suggest accepting the schemafactory in the JSON and default to JSON schemafactory

File: jdbc/src/main/java/com/qubole/quark/jdbc/schema/dbschema/DbSchemaFactory.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      long dsSetId = dsSets.get(0).getId();
+      long defaultDataSourceId = dsSets.get(0).getDefaultDatasourceId();
+      List<JdbcSource> jdbcSources = jdbcSourceDAO.findByDSSetId(dsSetId);
+      List<QuboleDbSource> quboleDbSources = quboleDbSourceDAO.findByDSSetId(dsSetId);
+
+      ImmutableList.Builder<com.qubole.quark.planner.DataSourceSchema> schemaList =
+          new ImmutableList.Builder<>();
+
+      for (DataSource dataSource : jdbcSources) {
+        schemaList.add(new DataSourceSchema(dataSource.getProperties(defaultDataSourceId)));
```
Comment (vrajat) [inline jdbc/src/main/java/com/qubole/quark/jdbc/schema/dbschema/DbSchemaFactory.java]: I understand why you are passing `defaultDataSourceId`. Not really your fault. It stinks though.

**Comment by vrajat** [comment]
You also need to add end-to-end tests.

**Comment by amoghmargoor** [comment]
I though only one migration would be present with the latest schema and new migrations would be added on top of that. Any reason why so many migrations are required in this change ?

**Comment by vrajat** [comment]
Pro for keeping all the migration files is that it is compatible with the internal system and we can merge them eventually. It is worth a debate.

File: jdbc/src/main/java/com/qubole/quark/jdbc/schema/dbschema/dao/DimensionDAO.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * Created by dev on 2/12/16.
+ */
+@RegisterMapper(DimensionMapper.class)
+public interface DimensionDAO {
+  @SqlQuery("select name, schema_name, table_name, column_name, "
+      + "cube_column_name, dimension_order, parent_dimension as "
+      + "parent  from dimensions where cube_id = :cube_id")
+  List<RelSchema.DbDimension> findByDSSetId(@Bind("cube_id") long cubeId);
```
Comment (amoghmargoor) [inline jdbc/src/main/java/com/qubole/quark/jdbc/schema/dbschema/dao/DimensionDAO.java]: name should change ?

File: jdbc/src/main/java/com/qubole/quark/jdbc/schema/dbschema/dao/MeasureDAO.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.List;
+
+/**
+ * Created by dev on 2/12/16.
+ */
+@RegisterMapper(MeasureMapper.class)
+public interface MeasureDAO {
+  @SqlQuery("select id, column_name, function, cube_column_name from "
+      + "measures where cube_id = :cube_id")
+  List<RelSchema.DbMeasure> findByDSSetId(@Bind("cube_id") long cubeId);
```
Comment (amoghmargoor) [inline jdbc/src/main/java/com/qubole/quark/jdbc/schema/dbschema/dao/MeasureDAO.java]: name should change ?

File: jdbc/pom.xml (hunk 1 above)
Comment (adeshr) [inline jdbc/pom.xml]: Encryption can be done using `javax.persistence jar`(similar as hibernate).

**Comment by amoghmargoor** [comment]
@vrajat  that is a good point to keep DB compatible with internal system. But for that keeping future migrations in sync with internal service is good enough. We do not require to carry old migrations for that.

**Comment by vrajat** [comment]
So lets say internally we are up to V10 internally. We could merge all the changes and call it V11. So V1-V10 == V11. Going forward we use V12 in both systems so that they can be merged eventually. Will that work ?

**Comment by amoghmargoor** [comment]
yeah... that should work. Our tests for nezha server currently runs migration V1 to V10. After merging I assume V1- V10 would be dropped requiring them to run just V11 onwards and that should be fine.

**Comment by vrajat** [comment]
I dont think we can have V11 in internal server. Current DBs have till V10. So internal server cannot be changed. Going forward, both can be in-sync from V12

File: jdbc/pom.xml [hunk 6]
```diff
@@ -105,6 +105,22 @@
...
+            <version>2.71</version>
+        </dependency>
+        <dependency>
+            <groupId>org.flywaydb</groupId>
+            <artifactId>flyway-core</artifactId>
+            <version>3.2.1</version>
+        </dependency>
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
```
Comment (vrajat) [inline jdbc/pom.xml]: Do you need this dependency for encryption ? Can you see if it can be removed ?

Comment (adeshr) [inline jdbc/pom.xml]: It will be needed for unit tests.

**Comment by vrajat** [comment]
@adeshr The code seems almost in shape. I see three more tasks:
1. Need unit tests - talk to @amoghmargoor or me to choose the tests.
2. Need to decide on flyway migrations. I am not convinced either way.
3. Module/Code organization. I'll start another comment about that.

**Comment by vrajat** [comment]
@adeshr @amoghmargoor I thinking about code organization.
This code seems like the catalog module of a database. This is what I propose:
- Start a new module catalog which contains schemafactories for both JSON and DB. Its the current code.
- Namespace is `com.qubole.quark.catalog.[json|db]`
- jdbc module adds it as a dependency.

The pros are that it is more usable if split into its own module.
- In the future, `quark-server` also adds it as a module.
- Possible a custom implementation of the server can also use it.

In the current state, both of them have to depend on `jdbc` module which is unnecessary.
What is your opinion ?

File: jdbc/pom.xml (hunk 6 above)
Comment (vrajat) [inline jdbc/pom.xml]: No. You should be using H2 for unit tests. Also add H2 in `<scope>test</scope>`

**Comment by vrajat** [comment]
@adeshr lets keep migrations as is.

**Comment by vrajat** [comment]
Logic & tests looks good. Code needs cleanup.
1. Can `pojo/*` and 'Relschema::Db*' classes be merged ?
2. There should be only one definition tests for `Select` and `Metadata`
3. Extract common code from model tests.

**Comment by adeshr** [comment]
Moving `pojo/Cube` to `RelSchema` can be confusing since there is already a `DbCube`.
`DbCube` can not be merged with `Cube` because:
1) DbCube needs List<Dimension>/List<Measure>/List<Groups> to be present.
2) To read dimensions/measures from db we need cube id to be present.
one needs the other, that's why `Cube` (Intermediary class) was created, which stores cube from db, then read dimensions/measures and create DbCube from cube/dimensions/measures.

**Comment by vrajat** [comment]
Good to merge. @adeshr did you try to rebase ? If not successful, I'll squash merge

**Comment by vrajat** [comment]
hmm - Just realized a missing piece. You are assuming that there is a record for DSSet. Lets tackle that in a different task.

**Pull request merged**